File: plugins/claude-code-hermit/scripts/hermit-stop.py
#!/usr/bin/env python3
"""Graceful shutdown for hermit autonomous sessions.

Sends /session-close --shutdown to the running Claude instance before
killing the tmux session, ensuring a clean session report is generated.

Usage:
    python scripts/hermit-stop.py              # graceful shutdown
    python scripts/hermit-stop.py --force      # immediate kill
"""

import fcntl
import json
import subprocess
import sys
import time
from pathlib import Path

CONFIG_PATH = Path('.claude-code-hermit/config.json')
STATE_DIR = CONFIG_PATH.parent / 'state'
RUNTIME_JSON = STATE_DIR / 'runtime.json'
RUNTIME_TMP = STATE_DIR / '.runtime.json.tmp'
LIFECYCLE_LOCK = STATE_DIR / '.lifecycle.lock'
SESSIONS_DIR = Path('.claude-code-hermit/sessions')
SHELL_PATH = SESSIONS_DIR / 'SHELL.md'
DEFAULT_TIMEOUT = 60  # seconds to wait for graceful close


def load_config():
    """Load config.json for session name."""
    if not CONFIG_PATH.exists():
        print('[hermit] No config found. Is this a hermit project?')
        sys.exit(1)

    with open(CONFIG_PATH) as f:
        return json.load(f)


def get_session_name(config):
    """Resolve tmux session name from config."""
    name = config.get('tmux_session_name', 'hermit-{project_name}')
    project_name = Path.cwd().name
    return name.replace('{project_name}', project_name)


def session_exists(name):
    """Check if tmux session exists."""
    result = subprocess.run(
        ['tmux', 'has-session', '-t', name],
        capture_output=True,
    )
    return result.returncode == 0


def save_config(config):
    """Write config back to disk."""
    try:
        with open(CONFIG_PATH, 'w') as f:
            json.dump(config, f, indent=2)
            f.write('\n')
    except OSError:
        pass


def write_runtime_json(data):
    """Atomic write to state/runtime.json."""
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    data['updated_at'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
    RUNTIME_TMP.write_text(json.dumps(data, indent=2) + '\n')
    RUNTIME_TMP.rename(RUNTIME_JSON)


def read_runtime_json():
    """Read state/runtime.json, return None if missing or invalid."""
    try:
        with open(RUNTIME_JSON) as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return None


def update_runtime_field(updates):
    """Read-modify-write runtime.json with atomic write."""
    runtime = read_runtime_json() or {}
    runtime.update(updates)
    write_runtime_json(runtime)


def acquire_lifecycle_lock():
    """Acquire exclusive lifecycle lock. Returns fd or exits on contention."""
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    lock_fd = open(LIFECYCLE_LOCK, 'w')
    try:
        fcntl.flock(lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except BlockingIOError:
        print('[hermit] Another lifecycle operation in progress. Aborting.')
        sys.exit(1)
    return lock_fd


def release_lifecycle_lock(lock_fd):
    """Release the lifecycle lock."""
    if lock_fd:
        try:
            fcntl.flock(lock_fd, fcntl.LOCK_UN)
            lock_fd.close()
        except OSError:
            pass


def main():
    force = '--force' in sys.argv

    config = load_config()
    lock_fd = acquire_lifecycle_lock()
    session_name = get_session_name(config)

    if not session_exists(session_name):
        runtime = read_runtime_json()
        if runtime and runtime.get('runtime_mode') == 'interactive':
            # Claude is still running in the operator's terminal — don't corrupt
            # lifecycle truth. The Stop hook (triggered when Claude exits) owns
            # the transition to idle.
            print('[hermit] Hermit is running in interactive mode.')
            print('[hermit] Terminate the Claude process in your terminal (Ctrl+C).')
            config['always_on'] = False
            save_config(config)
            release_lifecycle_lock(lock_fd)
            sys.exit(0)
        print(f'[hermit] No running session: {session_name}')
        config['always_on'] = False
        save_config(config)
        update_runtime_field({
            'session_state': 'idle',
            'shutdown_completed_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
        })
        sys.exit(0)

    if force:
        print(f'[hermit] Force-killing session: {session_name}')
        config['always_on'] = False
        save_config(config)
        subprocess.run(['tmux', 'kill-session', '-t', session_name])
        now = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
        update_runtime_field({
            'session_state': 'idle',
            'shutdown_requested_at': now,
            'shutdown_completed_at': now,
        })
        print('[hermit] Warning: session was not closed gracefully. SHELL.md may be stale.')
        return

    # Stop heartbeat first (only if enabled in config)
    if config.get('heartbeat', {}).get('enabled', False) and session_exists(session_name):
        print('[hermit] Stopping heartbeat...')
        subprocess.run([
            'tmux', 'send-keys', '-t', session_name,
            '/claude-code-hermit:heartbeat stop', 'Enter',
        ])
        time.sleep(2)

    # Mark shutdown requested in runtime.json
    update_runtime_field({
        'shutdown_requested_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
    })

    # Release the lifecycle lock before delegating to /done.
    # The done path inside Claude needs to acquire this lock
    # for its own runtime.json writes. Holding it here would cause the
    # agent to see it as contention and skip the close.
    release_lifecycle_lock(lock_fd)
    lock_fd = None

    # Graceful shutdown: send /done --shutdown for full close
    print(f'[hermit] Sending /claude-code-hermit:done --shutdown to {session_name}...')
    subprocess.run([
        'tmux', 'send-keys', '-t', session_name,
        '/claude-code-hermit:done --shutdown', 'Enter',
    ])

    # Wait for /done to mark shutdown complete in runtime.json. Poll
    # shutdown_completed_at >= shutdown_requested_at (UTC-Z lexical compare
    # is safe since Step 3 standardized all writers).
    print(f'[hermit] Waiting up to {DEFAULT_TIMEOUT}s for /done to complete shutdown...')
    closed = False
    for i in range(DEFAULT_TIMEOUT):
        time.sleep(1)
        if not session_exists(session_name):
            print('[hermit] Session exited.')
            closed = True
            break
        runtime = read_runtime_json() or {}
        done_at = runtime.get('shutdown_completed_at')
        req_at = runtime.get('shutdown_requested_at')
        if done_at and (not req_at or done_at >= req_at):
            print(f'[hermit] /done completed after {i}s.')
            closed = True
            break
    if not closed:
        print(f'[hermit] Timeout after {DEFAULT_TIMEOUT}s. Killing session.')

    # Re-acquire lock for final state writes and cleanup
    lock_fd = acquire_lifecycle_lock()

    # Reset always_on flag
    config['always_on'] = False
    save_config(config)

    # Kill tmux session
    if session_exists(session_name):
        subprocess.run(['tmux', 'kill-session', '-t', session_name])
        print(f'[hermit] tmux session "{session_name}" terminated.')

    # Mark shutdown completed in runtime.json. The /done path normally does this;
    # we set it defensively in case /done didn't fire (e.g., timeout).
    update_runtime_field({
        'session_state': 'idle',
        'shutdown_completed_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
    })


if __name__ == '__main__':
    main()

File: plugins/claude-code-hermit/scripts/hermit-stop.ts
#!/usr/bin/env bun
/**
 * Graceful shutdown for hermit autonomous sessions.
 *
 * Sends /session-close --shutdown to the running Claude instance before
 * killing the tmux session, ensuring a clean session report is generated.
 *
 * Usage:
 *     bun scripts/hermit-stop.ts              # graceful shutdown
 *     bun scripts/hermit-stop.ts --force      # immediate kill
 */

import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { acquireLock, releaseLock } from './lib/lockfile';
import { localISOStamp } from './lib/time';
import { readRuntimeJson, updateRuntimeField, STATE_DIR, LIFECYCLE_LOCK } from './lib/runtime';

type Json = any;

const CONFIG_PATH = '.claude-code-hermit/config.json';
const SESSIONS_DIR = '.claude-code-hermit/sessions';
const SHELL_PATH = path.join(SESSIONS_DIR, 'SHELL.md');
const DEFAULT_TIMEOUT = 60; // seconds to wait for graceful close

const sleep = (s: number) => new Promise((r) => setTimeout(r, s * 1000));

function loadConfig(): Json {
  if (!fs.existsSync(CONFIG_PATH)) {
    console.log('[hermit] No config found. Is this a hermit project?');
    process.exit(1);
  }
  return JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf-8'));
}

function getSessionName(config: Json): string {
  const name = config.tmux_session_name ?? 'hermit-{project_name}';
  return name.replaceAll('{project_name}', path.basename(process.cwd()));
}

function tmuxSessionAlive(name: string): boolean {
  return spawnSync('tmux', ['has-session', '-t', name], { stdio: 'ignore' }).status === 0;
}

function findLatestReport(): string | null {
  try {
    const reports = fs
      .readdirSync(SESSIONS_DIR)
      .filter((f) => /^S-.*-REPORT\.md$/.test(f))
      .sort();
    return reports.length ? path.join(SESSIONS_DIR, reports[reports.length - 1]) : null;
  } catch {
    return null;
  }
}

function listReports(): Set<string> {
  try {
    return new Set(fs.readdirSync(SESSIONS_DIR).filter((f) => /^S-.*-REPORT\.md$/.test(f)));
  } catch {
    return new Set();
  }
}

function readActiveSession(): Json | null {
  const stats: Json = {};

  // Status and started come from runtime.json — the documented single source of truth.
  // SHELL.md has no **Status:** field (removed; see session-mgr.md).
  const runtime = readRuntimeJson();
  if (runtime) {
    stats.status = runtime.session_state ?? 'unknown';
    stats.started = runtime.created_at ?? 'unknown';
  }

  // Prefer SHELL.md's **Started:** over runtime.json's only when it has been substituted
  // (i.e. does not still contain the template placeholder YYYY-MM-DD HH:MM).
  if (fs.existsSync(SHELL_PATH)) {
    for (const line of fs.readFileSync(SHELL_PATH, 'utf-8').split('\n')) {
      if (line.includes('**Tasks Completed:**')) {
        stats.tasks_completed = line.split('**Tasks Completed:**')[1].trim();
      } else if (line.includes('**Started:**')) {
        const value = line.split('**Started:**')[1].trim();
        if (!value.includes('YYYY')) stats.started = value;
      }
    }
  }

  return Object.keys(stats).length ? stats : null;
}

function saveConfig(config: Json): void {
  try {
    fs.writeFileSync(CONFIG_PATH, JSON.stringify(config, null, 2) + '\n');
  } catch {}
}

function acquireLifecycleLock(): void {
  fs.mkdirSync(STATE_DIR, { recursive: true });
  if (!acquireLock(LIFECYCLE_LOCK)) {
    console.log('[hermit] Another lifecycle operation in progress. Aborting.');
    process.exit(1);
  }
}

function releaseLifecycleLock(): void {
  releaseLock(LIFECYCLE_LOCK);
}

function tmux(args: string[]): void {
  spawnSync('tmux', args, { stdio: 'inherit' });
}

async function main(): Promise<void> {
  const force = process.argv.includes('--force');

  const config = loadConfig();
  acquireLifecycleLock();
  const sessionName = getSessionName(config);

  if (!tmuxSessionAlive(sessionName)) {
    const runtime = readRuntimeJson();
    if (runtime && runtime.runtime_mode === 'interactive') {
      // Claude is still running in the operator's terminal — don't corrupt
      // lifecycle truth. The Stop hook (triggered when Claude exits) owns
      // the transition to idle.
      console.log('[hermit] Hermit is running in interactive mode.');
      console.log('[hermit] Terminate the Claude process in your terminal (Ctrl+C).');
      config.always_on = false;
      saveConfig(config);
      releaseLifecycleLock();
      process.exit(0);
    }
    console.log(`[hermit] No running session: ${sessionName}`);
    config.always_on = false;
    saveConfig(config);
    updateRuntimeField({
      session_state: 'idle',
      shutdown_completed_at: localISOStamp(),
      transition: null,
      transition_target: null,
      transition_started_at: null,
    });
    const report = findLatestReport();
    if (report) console.log(`[hermit] Last report: ${report}`);
    releaseLifecycleLock();
    process.exit(0);
  }

  // Show session stats
  const stats = readActiveSession();
  let tasks = '0';
  if (stats) {
    tasks = stats.tasks_completed ?? '0';
    const started = stats.started ?? 'unknown';
    const status = stats.status ?? 'unknown';
    console.log(`[hermit] Session started: ${started} | Status: ${status} | Tasks: ${tasks}`);
  }

  if (force) {
    console.log(`[hermit] Force-killing session: ${sessionName}`);
    config.always_on = false;
    saveConfig(config);
    tmux(['kill-session', '-t', sessionName]);
    updateRuntimeField({
      session_state: 'idle',
      shutdown_requested_at: localISOStamp(),
      shutdown_completed_at: localISOStamp(),
      last_error: 'unclean_shutdown',
      transition: null,
      transition_target: null,
      transition_started_at: null,
    });
    const report = findLatestReport();
    if (report) console.log(`[hermit] Last report: ${report}`);
    console.log('[hermit] Warning: session was not closed gracefully. SHELL.md may be stale.');
    releaseLifecycleLock();
    return;
  }

  // Stop heartbeat first (only if enabled in config)
  if (config.heartbeat?.enabled && tmuxSessionAlive(sessionName)) {
    console.log('[hermit] Stopping heartbeat...');
    tmux(['send-keys', '-t', sessionName, '/claude-code-hermit:heartbeat stop', 'Enter']);
    await sleep(2);
  }

  // Mark shutdown requested in runtime.json
  updateRuntimeField({ shutdown_requested_at: localISOStamp() });

  // Release the lifecycle lock before delegating to /session-close.
  // The close/archive path inside Claude needs to acquire this lock
  // for its own runtime.json writes. Holding it here would cause the
  // agent to see it as contention and skip the close.
  releaseLifecycleLock();

  // Graceful shutdown: send /session-close --shutdown for full close
  console.log(`[hermit] Sending /claude-code-hermit:session-close --shutdown to ${sessionName}...`);
  tmux(['send-keys', '-t', sessionName, '/claude-code-hermit:session-close --shutdown', 'Enter']);

  // Wait for the session to close (check for new report file)
  const reportsBefore = listReports();
  console.log(`[hermit] Waiting up to ${DEFAULT_TIMEOUT}s for session close...`);

  let newReport: string | null = null;
  let closedEarly = false;
  for (let i = 0; i < DEFAULT_TIMEOUT; i++) {
    await sleep(1);
    if (!tmuxSessionAlive(sessionName)) {
      console.log('[hermit] Session exited without generating a report.');
      closedEarly = true;
      break;
    }
    const fresh = [...listReports()].filter((r) => !reportsBefore.has(r));
    if (fresh.length) {
      newReport = path.join(SESSIONS_DIR, fresh[0]);
      console.log(`[hermit] Session closed. Report: ${newReport}`);
      closedEarly = true;
      break;
    }
  }
  if (!closedEarly) {
    console.log(`[hermit] Timeout after ${DEFAULT_TIMEOUT}s. Killing session.`);
  }

  // Re-acquire lock for final state writes and cleanup
  acquireLifecycleLock();

  // Reset always_on flag
  config.always_on = false;
  saveConfig(config);

  // Kill tmux session
  if (tmuxSessionAlive(sessionName)) {
    tmux(['kill-session', '-t', sessionName]);
    console.log(`[hermit] tmux session "${sessionName}" terminated.`);
  }

  // Mark shutdown completed in runtime.json
  const shutdownUpdates: Json = {
    session_state: 'idle',
    shutdown_completed_at: localISOStamp(),
    transition: null,
    transition_target: null,
    transition_started_at: null,
  };
  if (!newReport) shutdownUpdates.last_error = 'unclean_shutdown';
  updateRuntimeField(shutdownUpdates);

  // Show summary
  if (!newReport) {
    const report = findLatestReport();
    if (report) console.log(`[hermit] Latest report: ${report}`);
  }
  if (stats) console.log(`[hermit] Total tasks this session: ${tasks}`);

  releaseLifecycleLock();
}

if (import.meta.main) {
  await main();
}
